### PR TITLE
WFLY-17322 Clustering subsystems create duplicate PersistentResourceXMLDescription instances

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.redhat.com/browse/WFLY-17322

I recommend reviewing individual commits.

### CODE REVIEW

File: clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    private final Model currentModel;
+    private final Supplier<ManagementRegistrar<SubsystemRegistration>> registrarFactory;
+    private final S currentSchema;
+    private final Function<S, XMLElementReader<List<ModelNode>>> readerFactory;
+    private final XMLElementWriter<SubsystemMarshallingContext> writer;
+    private final XMLElementReader<List<ModelNode>> currentReader;
+
+    /**
+     * Constructs a new extension using a reader/writer factory.
+     * @param <P> the parser type
```
Comment (rhusar) [inline clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java:60]:
```suggestion
     * @param <S> the parser type
```

Though, unfortunately Javadoc doesn't allow for that at all here.

```
Generating /tmp/org/jboss/as/clustering/controller/SubsystemExtension.html...
/Users/rhusar/git/wildfly/clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java:60: error: @param name not found
     * @param <S> the parser type
               ^
```

File: clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    private SubsystemExtension(String name, Model currentModel, Supplier<ManagementRegistrar<SubsystemRegistration>> registrarFactory, S currentSchema, Function<S, PersistentResourceXMLDescription> descriptionFactory, PersistentResourceXMLDescription currentDescription) {
+        // Reuse current xml description between reader and writer
+        this(name, currentModel, registrarFactory, currentSchema, descriptionFactory.andThen(XMLDescriptionReader::new), new XMLDescriptionWriter(currentDescription), new XMLDescriptionReader(currentDescription));
+    }
+
+    /**
+     * Constructs a new extension using a reader factory and a separate writer implementation.
+     * @param <P> the parser type
```
Comment (rhusar) [inline clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java:79]:
```suggestion
     * @param <S> the parser type
```

Though, unfortunately Javadoc doesn't allow for that at all here.

```
Generating /tmp/org/jboss/as/clustering/controller/SubsystemExtension.html...
/Users/rhusar/git/wildfly/clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java:60: error: @param name not found
     * @param <S> the parser type
               ^
```

**Review by rhusar — APPROVED** [review]
Changes look good to me; just Javadoc nitpicks.

**Comment by pferraro** [comment]
Thanks - I'll amend the javadoc fixes into the appropriate commit and merge when CI completes.

File: clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java (hunk 1 above)
Comment (pferraro) [reply clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java:60]: Hmm - that would be a bug in the eclipse javadoc auto-generation.

**Pull request merged**